Clarity required over NRT approach for Data Transfer Objects.

As requested by @BillWagner on twitter, I'm raising an issue around Nullable Reference Types and how they apply to Data Transfer Objects (DTOs).

Up until the introduction of NRTs, Data Transfer Objects have often not had an explicit constructor:

```csharp
class Student 
{
        [Required]
        public string FirstName { get; set; }

        [Required]
        public string LastName { get; set; }

        public string OptionalField { get; set; }
}
```

Mandatory fields (that cannot be null) were marked with ``[Required]`` or a similar appropriate attribute.

When we introduce NRTs, things get a bit more complicated as we attempt to use NRTs, where they overlap with what are effectively schema/business constraints.

```csharp
using System.ComponentModel.DataAnnotations;

namespace NullDto
{
    // Example Data Transfer Object - *not* specific to any ORM.
    //
    // From the perspective of the application, the [Required] fields are not optional, but are often 
    // populated automatically by a deserialiser. So there is an empty constructor to allow for that, which
    // results in the below warnings.
    class Student 
    {
        // warning CS8618: Non-nullable property 'FirstName' is uninitialized. Consider declaring the property as nullable.
        // warning CS8618: Non-nullable property 'LastName' is uninitialized. Consider declaring the property as nullable.
        public Student()
        {
            // Empty constructor used for deserialisation. 
        }

        public Student(string firstName, string lastName)
        {            
            // Constructor used when creating a new Student in code.
            FirstName = firstName;
            LastName = lastName;
        }

        [Required]
        public string FirstName { get; set; }

        [Required]
        public string LastName { get; set; }

        // No warning here, this field is actually optional (note no [Required] attribute).
        public string? OptionalField { get; set; }
    }
}
```

I could simply indicate that ``FirstName`` and ``LastName`` may in fact be null, but that's only true from a **language** standpoint; in practice, they will never be null, and I don't want to add to the cognitive load for developers by forcing them to check for null on properties that in practice are never null.

Current Options:

- Disable nullable behaviour in that one file (don't like that).
- Initialise ``FirstName`` and ``LastName`` to some default value (like ``String.Empty``) in the default constructor. Not a fan either; might work for strings, but becomes very complicated for any more interesting types.
- Suppress the warnings on the constructor. Again, not ideal, but at least it gives us an opportunity to document why we've done it.

Possible Solutions:

- Decorate a property with an attribute, like ``[ExternallyInitialized]`` or something similar, which indicates that a property is set externally. This could also mean that instantiating an instance through property initialization would generate warnings if you *didn't* set that property. Might have some overlap with the planned ``init`` access modifier?
- Treat classes differently if they have ``[Serializable]`` on the type (or something like that).

Anything else? I don't know what the best practice is here, or if I've missed something major?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a690704d-da32-dca7-ea3e-66b84fb1d00d
* Version Independent ID: 11f314de-dd3a-362e-b2bb-d0a2133cfc7b
* Content: [Nullable reference types](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references)
* Content Source: [docs/csharp/nullable-references.md](https://github.com/dotnet/docs/blob/master/docs/csharp/nullable-references.md)
* Product: **dotnet-csharp**
* Technology: **csharp-null-safety**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

--- a/docs/csharp/nullable-migration-strategies.md
+++ b/docs/csharp/nullable-migration-strategies.md
@@ -179,7 +179,7 @@ public string FirstName
 }
 ```
 
-In this scenario, if the `FirstName` property is accessed before it has been initialized, then we throw an `InvalidOperationException`, because the API contract has been used incorrectly.
+In this scenario, if the `FirstName` property is accessed before it has been initialized, then the code throws an `InvalidOperationException`, because the API contract has been used incorrectly.
 
 You should consider that some libraries may have special considerations when using backing fields. For example, EF Core may need to be configured to use [backing fields](/ef/core/modeling/backing-field) correctly.
 
@@ -197,7 +197,7 @@ public string LastName { get; set; } = null!;
 public string? VehicleRegistration { get; set; }
 ```
 
-You will never observe an actual null value at runtime except as a result of a programming bug (i.e. by accessing the property before it has been properly initialized).
+You will never observe an actual null value at runtime except as a result of a programming bug, by accessing the property before it has been properly initialized.
 
 ## See also
 

--- a/docs/csharp/nullable-migration-strategies.md
+++ b/docs/csharp/nullable-migration-strategies.md
@@ -89,7 +89,7 @@ You may want to restrict the types used for a generic type argument to be non-nu
 
 ## Late-initialized properties, Data Transfer Objects and nullability
 
-Indicating the nullability of properties that are late-initialized (i.e. set after construction) may require special consideration to ensure that your class continues to correctly express the original design intent.
+Indicating the nullability of properties that are late-initialized, meaning set after construction, may require special consideration to ensure that your class continues to correctly express the original design intent.
 
 Types that contain late-initialized properties, such as Data Transfer Objects (DTOs), are often instantiated by an external library, like a database ORM (Object Relational Mapper), a deserializer, or some other component that automatically populates properties from another source.
 
@@ -108,11 +108,11 @@ class Student
 }
 ```
 
-The design intent (indicated in this case by the ``Required`` attribute) suggests that in this system, the ``FirstName`` and ``LastName`` properties are **mandatory**, and therefore not null.
+The design intent (indicated in this case by the `Required` attribute) suggests that in this system, the `FirstName` and `LastName` properties are **mandatory**, and therefore not null.
 
-The ``VehicleRegistration`` property is **not mandatory**, so may be null.
+The `VehicleRegistration` property is **not mandatory**, so may be null.
 
-When we enable nullable reference types, we want to indicate on our DTO which of the properties may be nullable, consistent with our original intent:
+When you enable nullable reference types, you want to indicate on our DTO which of the properties may be nullable, consistent with your original intent:
 
 ```csharp
 class Student
@@ -129,9 +129,9 @@ class Student
 
 For this DTO, the only property that may be null is ``VehicleRegistration``.
 
-However, the compiler raises ``CS8618`` warnings for both ``FirstName`` and ``LastName``, indicating the non-nullable properties are uninitialized.
+However, the compiler raises `CS8618` warnings for both `FirstName` and `LastName`, indicating the non-nullable properties are uninitialized.
 
-There are three options available to you that resolve the compiler warnings, in a way that maintains the original intent. Any of these options are valid; you should choose the one that best suits your coding style and design requirements.
+There are three options available to you that resolve the compiler warnings in a way that maintains the original intent. Any of these options are valid; you should choose the one that best suits your coding style and design requirements.
 
 ### Initialize in the constructor
 
@@ -179,13 +179,13 @@ public string FirstName
 }
 ```
 
-In this scenario, if the ``FirstName`` property is accessed before it has been initialized, then we throw an ``InvalidOperationException``, because the API contract has been used incorrectly.
+In this scenario, if the `FirstName` property is accessed before it has been initialized, then we throw an `InvalidOperationException`, because the API contract has been used incorrectly.
 
 You should consider that some libraries may have special considerations when using backing fields. For example, EF Core may need to be configured to use [backing fields](/ef/core/modeling/backing-field) correctly.
 
 ### Initialize the property to null
 
-As a terser alternative to using a nullable backing field, or if the library that instantiates your class is not compatible with that approach, you can simply initialize the property to null directly, with the help of the null-forgiving operator (``!``):
+As a terser alternative to using a nullable backing field, or if the library that instantiates your class is not compatible with that approach, you can simply initialize the property to `null` directly, with the help of the null-forgiving operator (`!`):
 
 ```csharp
 [Required]

--- a/docs/csharp/nullable-migration-strategies.md
+++ b/docs/csharp/nullable-migration-strategies.md
@@ -87,11 +87,11 @@ What it does mean is that you can't use `T?` in a generic class or method declar
 
 You may want to restrict the types used for a generic type argument to be non-nullable types. You can do that by adding the `notnull` constraint on that type argument. When that constraint is applied, the type argument must not be a nullable type.
 
-## Updating Data Transfer Objects (DTOs) to indicate nullability
+## Late-initialized properties, Data Transfer Objects and nullability
 
-Indicating the nullability of the properties of Data Transfer Objects (DTOs) in your code-base requires special consideration, to ensure that your class continues to correctly express the original design intent of the DTO.
+Indicating the nullability of properties that are late-initialized (i.e. set after construction) may require special consideration to ensure that your class continues to correctly express the original design intent.
 
-DTOs are often instantiated by an external library, like a database ORM (Object Relational Mapper), a deserializer, or some other component that automatically populates properties from another source.
+Types that contain late-initialized properties, such as Data Transfer Objects (DTOs), are often instantiated by an external library, like a database ORM (Object Relational Mapper), a deserializer, or some other component that automatically populates properties from another source.
 
 Consider the following DTO class, prior to enabling nullable reference types, that represents a student:
 
@@ -161,7 +161,7 @@ This approach only works if the library that you use to instantiate the class su
 In addition, a library may support passing *some* properties in the constructor, but not all.
 For example, EF Core supports [constructor binding](/ef/core/modeling/constructors) for normal column properties, but not navigation properties.
 
-Check the documentation on the library that instantiates the DTO, to understand the extent to which it supports constructor binding.
+Check the documentation on the library that instantiates your class, to understand the extent to which it supports constructor binding.
 
 ### Property with nullable backing field
 
@@ -185,7 +185,7 @@ You should consider that some libraries may have special considerations when usi
 
 ### Initialize the property to null
 
-As a terser alternative to using a nullable backing field, or if the library that instantiates your DTO is not compatible with that approach, you can simply initialize the property to null directly, with the help of the null-forgiving operator (``!``):
+As a terser alternative to using a nullable backing field, or if the library that instantiates your class is not compatible with that approach, you can simply initialize the property to null directly, with the help of the null-forgiving operator (``!``):
 
 ```csharp
 [Required]

--- a/docs/csharp/nullable-migration-strategies.md
+++ b/docs/csharp/nullable-migration-strategies.md
@@ -86,3 +86,120 @@ This doesn't mean you can't use a nullable type (either value type or reference 
 What it does mean is that you can't use `T?` in a generic class or method declaration without constraints. For example, <xref:System.Linq.Enumerable.FirstOrDefault%60%601(System.Collections.Generic.IEnumerable%7B%60%600%7D)?displayProperty=nameWithType> won't be changed to return `T?`. You can overcome this limitation by adding either the `struct` or `class` constraint. With either of those constraints, the compiler knows how to generate code for both `T` and `T?`.
 
 You may want to restrict the types used for a generic type argument to be non-nullable types. You can do that by adding the `notnull` constraint on that type argument. When that constraint is applied, the type argument must not be a nullable type.
+
+## Updating Data Transfer Objects (DTOs) to indicate nullability
+
+Indicating the nullability of the properties of Data Transfer Objects (DTOs) in your code-base requires special consideration, to ensure that your class continues to correctly express the original design intent of the DTO.
+
+DTOs are often instantiated by an external library, like a database ORM (Object Relational Mapper), a deserializer, or some other component that automatically populates properties from another source.
+
+Consider the following DTO class, prior to enabling nullable reference types, that represents a student:
+
+```csharp
+class Student
+{
+    [Required]
+    public string FirstName { get; set; }
+
+    [Required]
+    public string LastName { get; set; }
+
+    public string VehicleRegistration { get; set; }
+}
+```
+
+The design intent (indicated in this case by the ``Required`` attribute) suggests that in this system, the ``FirstName`` and ``LastName`` properties are **mandatory**, and therefore not null.
+
+The ``VehicleRegistration`` property is **not mandatory**, so may be null.
+
+When we enable nullable reference types, we want to indicate on our DTO which of the properties may be nullable, consistent with our original intent:
+
+```csharp
+class Student
+{
+    [Required]
+    public string FirstName { get; set; }
+
+    [Required]
+    public string LastName { get; set; }
+
+    public string? VehicleRegistration { get; set; }
+}
+```
+
+For this DTO, the only property that may be null is ``VehicleRegistration``.
+
+However, the compiler raises ``CS8618`` warnings for both ``FirstName`` and ``LastName``, indicating the non-nullable properties are uninitialized.
+
+There are three options available to you that resolve the compiler warnings, in a way that maintains the original intent. Any of these options are valid; you should choose the one that best suits your coding style and design requirements.
+
+### Initialize in the constructor
+
+The ideal way to resolve the uninitialized warnings is to initialize the properties in the constructor:
+
+```csharp
+class Student
+{
+    public Student(string firstName, string lastName)
+    {
+        FirstName = firstName;
+        LastName = lastName;
+    }
+
+    [Required]
+    public string FirstName { get; set; }
+
+    [Required]
+    public string LastName { get; set; }
+
+    public string? VehicleRegistration { get; set; }
+}
+```
+
+This approach only works if the library that you use to instantiate the class supports passing parameters in the constructor.
+
+In addition, a library may support passing *some* properties in the constructor, but not all.
+For example, EF Core supports [constructor binding](/ef/core/modeling/constructors) for normal column properties, but not navigation properties.
+
+Check the documentation on the library that instantiates the DTO, to understand the extent to which it supports constructor binding.
+
+### Property with nullable backing field
+
+If constructor binding won't work for you, one way to deal with this problem is to have a non-nullable property with a nullable backing field:
+
+```csharp
+private string? _firstName;
+
+[Required]
+public string FirstName
+{
+    set => _firstName = value;
+    get => _firstName
+           ?? throw new InvalidOperationException("Uninitialized " + nameof(FirstName))
+}
+```
+
+In this scenario, if the ``FirstName`` property is accessed before it has been initialized, then we throw an ``InvalidOperationException``, because the API contract has been used incorrectly.
+
+You should consider that some libraries may have special considerations when using backing fields. For example, EF Core must be configured to use [backing fields](/ef/core/modeling/backing-field) correctly.
+
+### Initialize the property to null
+
+As a terser alternative to using a nullable backing field, or if the library that instantiates your DTO is not compatible with that approach, you can simply initialize the property to null directly, with the help of the null-forgiving operator (``!``):
+
+```csharp
+[Required]
+public string FirstName { get; set; } = null!;
+
+[Required]
+public string LastName { get; set; } = null!;
+
+public string? VehicleRegistration { get; set; }
+```
+
+You will never observe an actual null value at runtime except as a result of a programming bug (e.g. by accessing the property before it has been properly initialized).
+
+## See also
+
+- [Migrate an existing codebase to nullable references](tutorials/upgrade-to-nullable-references.md)
+- [Working with Nullable Reference Types in EF Core](/ef/core/miscellaneous/nullable-reference-types)

--- a/docs/csharp/nullable-migration-strategies.md
+++ b/docs/csharp/nullable-migration-strategies.md
@@ -181,7 +181,7 @@ public string FirstName
 
 In this scenario, if the ``FirstName`` property is accessed before it has been initialized, then we throw an ``InvalidOperationException``, because the API contract has been used incorrectly.
 
-You should consider that some libraries may have special considerations when using backing fields. For example, EF Core must be configured to use [backing fields](/ef/core/modeling/backing-field) correctly.
+You should consider that some libraries may have special considerations when using backing fields. For example, EF Core may need to be configured to use [backing fields](/ef/core/modeling/backing-field) correctly.
 
 ### Initialize the property to null
 
@@ -197,7 +197,7 @@ public string LastName { get; set; } = null!;
 public string? VehicleRegistration { get; set; }
 ```
 
-You will never observe an actual null value at runtime except as a result of a programming bug (e.g. by accessing the property before it has been properly initialized).
+You will never observe an actual null value at runtime except as a result of a programming bug (i.e. by accessing the property before it has been properly initialized).
 
 ## See also
 

